### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by guderkar**

### CODE REVIEW

**Review by ds-lukaslangr — CHANGES_REQUESTED** [review]

File: src/featurestorebundle/feature/FeatureList.py [hunk 1]
```diff
@@ -19,6 +19,13 @@ def empty(self):
     def get_names(self):
         return [feature.name for feature in self.__features]
 
+    def get_by_name(self, feature_name) -> FeatureInstance:
```
[the remark this answers is not in the document]
Comment (guderkar) [reply src/featurestorebundle/feature/FeatureList.py:22]: forgot to delete

File: src/featurestorebundle/feature/FeatureInstance.py [hunk 2]
```diff
@@ -40,6 +40,10 @@ def description(self):
     def dtype(self):
         return self.__dtype
 
+    @property
+    def storage_dtype(self):
+        return f"map<tinyint,{self.__dtype}>" if self.__template.default_value is None else self.__dtype
```
[the remark this answers is not in the document]
Comment (guderkar) [reply src/featurestorebundle/feature/FeatureInstance.py:45]: yep both work, byte is more feasible

File: src/featurestorebundle/delta/feature/FeaturesPreparer.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        self.__logger = logger
+        self.__features_joiner = features_joiner
+        self.__null_handler = null_handler
+        self.__checkpoint_guard = checkpoint_guard
+        self.__checkpoint_dir_setter = checkpoint_dir_setter
+
+    def prepare(self, features_storage: FeaturesStorage) -> DataFrame:
+        features_data = self.__features_joiner.join(features_storage)
+        features_data = self.__null_handler.to_storage_format(features_data, features_storage.feature_list)
+        features_data = self.__null_handler.fill_nulls(features_data, features_storage.feature_list)
```
[the remark this answers is not in the document]
Comment (guderkar) [reply src/featurestorebundle/delta/feature/FeaturesPreparer.py:28]: I think it does not matter but yes it makes more sense to have it swapped

File: src/featurestorebundle/delta/feature/NullHandler.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        fill_dict = {
+            feature.name: feature.template.default_value for feature in feature_list.get_all() if feature.template.default_value is not None
+        }
+
+        return df.fillna(fill_dict)
+
+    def to_storage_format(self, df: DataFrame, feature_list: FeatureList) -> DataFrame:
+        for feature in feature_list.get_all():
+            if feature.template.default_value is None:
+                df = df.withColumn(feature.name, f.create_map(f.lit(0), f.col(feature.name)).cast(feature.storage_dtype))
```
Comment (ds-lukaslangr) [inline src/featurestorebundle/delta/feature/NullHandler.py:20]: isn't it maybe better to make a list of Columns and then select them at once? this is slower according to Olda :)

File: src/featurestorebundle/delta/feature/NullHandler.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+        return df
+
+    def from_storage_format(self, df: DataFrame, feature_list: FeatureList) -> DataFrame:
+        for feature in feature_list.get_all():
+            if feature.template.default_value is None:
+                df = df.withColumn(feature.name, f.col(feature.name).getItem(0))
```
Comment (ds-lukaslangr) [inline src/featurestorebundle/delta/feature/NullHandler.py:27]: same as above

File: src/featurestorebundle/delta/feature/NullHandler.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+        return df
+
+    def __check_default_values_valid(self, feature_list: FeatureList):
+        for feature in feature_list.get_all():
+            if feature.template.default_value in [np.nan, np.inf]:
```
Comment (ds-lukaslangr) [inline src/featurestorebundle/delta/feature/NullHandler.py:33]: there is also negative infinity np.ninf, but I would rather use built-in float("nan"), float("inf") and float("-inf"), they are equivalent and we don't have to depend on numpy

File: src/featurestorebundle/delta/feature/FeaturesFilteringManager.py [hunk 7]
```diff
@@ -19,34 +25,25 @@ def get_latest(
...
-                    f.array_max(
-                        f.collect_list(f.struct(f.array_contains("features_intersect", feature), f.col(time_column), f.col(feature)))
-                    )[feature].alias(feature)
-                    for feature in relevant_features
-                ],
-                f.array_distinct(f.flatten(f.collect_set("features_intersect"))).alias("computed_features"),
-            )
-            .filter(f.size("computed_features") == len(relevant_features))
-            .drop("computed_features")
+            feature_store.select(id_column, *[f.first(feature, ignorenulls=True).over(window).alias(feature) for feature in features])
```
Comment (ds-lukaslangr) [inline src/featurestorebundle/delta/feature/FeaturesFilteringManager.py:33]: if you swap the comprehension from [] to () it should at least use much less memory compared to list

Comment (ds-lukaslangr) [reply src/featurestorebundle/delta/feature/FeaturesFilteringManager.py:33]: you could even make it a `map(lambda feature: f.first(feature, ignorenulls=True).over(window).alias(feature)), features)` to make it super fast

**Review by ds-lukaslangr — CHANGES_REQUESTED** [review]

File: src/featurestorebundle/delta/feature/NullHandler.py (hunk 4 above)
Comment (guderkar) [reply src/featurestorebundle/delta/feature/NullHandler.py:20]: True

File: src/featurestorebundle/delta/feature/FeaturesFilteringManager.py (hunk 7 above)
Comment (guderkar) [reply src/featurestorebundle/delta/feature/FeaturesFilteringManager.py:33]: I think that in the end it's same. So I would rather prefer list comprehension as it is more pythonic
image

File: src/featurestorebundle/delta/feature/NullHandler.py (hunk 6 above)
Comment (guderkar) [reply src/featurestorebundle/delta/feature/NullHandler.py:33]: are they really equivalent? Maybe after conversion to dataframe
image

Comment (guderkar) [reply src/featurestorebundle/delta/feature/NullHandler.py:33]: looks like we have to do this
```
if feature.template.default_value in [np.nan, np.inf, -np.inf, float("nan"), float("inf"), float("-inf")]:
```

Comment (guderkar) [reply src/featurestorebundle/delta/feature/NullHandler.py]: or `math.isnan(default_value) or math.isinf(default_value)`

**Pull request merged**